### ISSUE TITLE
Adding new modalities to the TerraMind model

### ISSUE THREAD

**Issue by aapowadi** [IBM/terratorch#754]
I wanted to customize the terramind pre-training with agriculture specific modalities but, I do not see any provisions to add new modalities to the pipeline. Also, I wanted to request to add documentation to pre-train the TerraMind model.

**Comment by Joao-L-S-Almeida** [comment]
@blumenstiel, any comment about this question ?

**Comment by blumenstiel** [comment]
Hi @aapowadi, we did not open source the pre-training code yet.

About adding new modalities in fine-tuning, you can define them like this:
```
backbone_modalities:
   - S2L2A  # Examples for pre-training modalities
   - {"NDVI": 1}  # Add new modality and define number of channels
```
This would add a new modality called "NDVI" with one channel and fine-tune the new patch-embedding (It's important to not freeze the encoder).

**Comment by Joao-L-S-Almeida** [comment]
Is the Benedikt's explanation solving your question @aapowadi ?

**Comment by aapowadi** [comment]
How would we train the tokenizer for this new modality ? Or is it not required?

**Comment by blumenstiel** [comment]
@aapowadi For fine-tuning, we don't use tokenizer to keep the model efficient. They are only used for generating pre-trianing targets. The fine-tuned model only uses the patch embeddings. If you define your modality as above, it wit automatically create a new patch embedding which is fine-tuned

**Comment by aapowadi** [comment]
I see. Can you also point me towards documentation/sample regression dataset. I don’t know how to format the regression dataset that I have to make it work with the terramind.

**Comment by blumenstiel** [comment]
Do you have a pixel-wise regression dataset or a image-level one? If it is pixel-wise, you can format it similarly to Sen1Floods11 or BurnScars and instead of saving the target class index, you save the regression values, see configs for details.

Not sure, if we have an example for image-level, @Joao-L-S-Almeida? But I can provide you some more explainations in that case.

**Comment by aapowadi** [comment]
It’s an image level regression dataset.

**Comment by blumenstiel** [comment]
In that case, you have two options: If you have some experience with torchgeo, you can simply create your own custom data module and just load it with the `"label"` key in the sample dict.

@Joao-L-S-Almeida if we have a example, please share.

Alternatively, you can also use the generic multimodal datamodule with `task: scalar_regression`. I implemented support from a scalar labels, but I did not test it very carefully because we don't work on such datasets. If you have any issues, let me know.
The docs are here: https://ibm.github.io/terratorch/stable/package/generic_datamodules/#terratorch.datamodules.generic_multimodal_data_module.GenericMultiModalDataModule

You need to create a csv/parquet file with the indexes being your sample file name ids and the the first column been your scalar label. You can provide the path to this file for the `{split}_label_data_root` arguments (all samples from all splits can be in a shared file or different ones, as you like). Next, you need to use `split` files that match your sample ids (e.g., that define the beginning of the name of your samples files, see Sen1Floods11 as an example). The code will then match these sample ids with the csv/parquet labels. 

It might require some debugging on your end. If you create the dataset, I would first test the generic dataest itself before using it for fine-tuning.

**Comment by aapowadi** [comment]
Thank you. Could you please help me with one more thing. I have a time-series dataset and I want to use all the time-series data to make predictions. How can I do that with this setup ?

**Comment by Joao-L-S-Almeida** [comment]
@aapowadi I recommend you to open a new issue or to create a question under "Discussions", maybe it's a better place to get support about tutorials.

**Comment by blumenstiel** [comment]
@aapowadi There is a temporal wrapper that applies an encoder to each step and combines the latents before the decoder. @Isabelle-Wittmann I could not find any docs or a example using it. Can you share one/add one to the docs?

**Comment by aapowadi** [comment]
Hello ! Thank you for pointing me to the temporal wrapper. That helped things along quite well. Could you also tell me if there is any method to save the finetuned latents of the terramind ?

**Comment by blumenstiel** [comment]
Hi @aapowadi, we have a embedding generation task in TerraTorch (in development).

See https://github.com/IBM/terratorch/blob/main/examples/confs/embedding_generation/terramind_embeddings.yaml
and https://ibm.github.io/terratorch/stable/tutorials/embedding_generation/?h=embedding.
@Isabelle-Wittmann is working on it if you have any questions.
